netsim/extra/evpn.multihoming/plugin.py

```diff
@@ -0,0 +1,206 @@
+        '_lacp_system_id': unique_ids[1],
+    }
+    return _auto_segments[esname]
+
+def _update_esi_stats(node: Box, interface: Box, esi_name: str, esi_data: dict) -> None:
+    if log.debug_active('plugin'):
+        print(f"evpn.multihoming: CALLED ESI STATS {node.name} {interface.ifname} {esi_name} {esi_data}")
+    # initialize struct if required
+    global _esi_stats
+    if esi_name not in _esi_stats.esi_reference: _esi_stats.esi_reference[esi_name] = []
+    if esi_name not in _esi_stats.esi_lacp_id: _esi_stats.esi_lacp_id[esi_name] = set()  # use set to avoid duplicates
+    if esi_name not in _esi_stats.esi_lacp_id_info: _esi_stats.esi_lacp_id_info[esi_name] = {}
+    
+    # increment usage data
+    _esi_stats.esi_reference[esi_name].append(f"{node.name}->{interface.ifname}")
+
+    # handle ESI-LAG case
+    if interface.type == 'lag' and interface.get('lag.lacp_system_id',False):
+        lacp_sys_id = interface.lag.lacp_system_id
+        _esi_stats.esi_lacp_id[esi_name].add(lacp_sys_id)
+        if lacp_sys_id in _esi_stats.esi_lacp_id_info[esi_name]:
+            # update
+            _esi_stats.esi_lacp_id_info[esi_name][lacp_sys_id].append(f"{node.name}->{interface.ifname}")
+        else:
+            # create
+            _esi_stats.esi_lacp_id_info[esi_name][lacp_sys_id] = [ f"{node.name}->{interface.ifname}" ]
+    return
+
+def _generate_esi_warnings() -> None:
+    global _esi_stats
+    # ESI used only once
+    for esi_name, esi_use in _esi_stats.esi_reference.items():
+        if len(esi_use) == 1:
+            log.warning(
+                text=f'EVPN Ethernet Segment "{esi_name}" seems to be used only once ({esi_use[0]})',
+                module=_config_name,
+                flag='evpn_mh_segment_used_once',
+                )
+    # ESI LAG LACP mismatch
+    for esi_name, esi_lacp_ids in _esi_stats.esi_lacp_id.items():
+        if len(esi_lacp_ids) != 1:
+            mismatch_list = []
+            for m in list(esi_lacp_ids):
+                found_interfaces=",".join(_esi_stats.esi_lacp_id_info[esi_name][m])
+                mismatch_list.append("[{}]->[{}]".format(m, found_interfaces))
+            mismatch_txt = ", ".join(mismatch_list)
+            log.warning(
+                text=f'EVPN ESI-LAG "{esi_name}" LACP mismatch. Found: {mismatch_txt}',
+                module=_config_name,
+                flag='evpn_mh_lacp_mismatch',
+                )
+    return
+
+def post_transform(topology: Box) -> None:
+    global _config_name
+    global _auto_segments
+    global _esi_stats
+    _init_es_id_generator(topology)
+    for node in topology.nodes.values():
+        if not 'evpn' in node.module: continue
+        features = devices.get_device_features(node,topology.defaults)
+        es_supported = 'evpn.multihoming' in features
+        if not es_supported: continue
+        # Load Ethernet Segments data and expand it with auto segments
+        es_data = node.get('evpn.ethernet_segments', {})
```

Comment:
Do you want to update this data structure in isolation or in-place? This will probably give you a pointer to **node.evpn.ethernet_segments** if it exists and an isolated dictionary otherwise.

I'd go with simple **es_data = node.evpn.ethernet_segments** if you want to do an update of the node data structure. This would create an empty dictionary within the node data and return a pointer to it (which would then be stored in es_data)